spec/models/ucas_matched_application_spec.rb

```diff
@@ -1,77 +1,70 @@
-      ucas_matching_data =
-        { 'Scheme' => 'U',
+    context 'when a course is not on Apply' do
+      it 'returns the correct course details' do
+        allow(Provider).to receive(:find_by).with(code: course.provider.code.to_s).and_return(course.provider)
+
+        ucas_matching_data = {
+          'Scheme' => 'U',
           'Course code' => '123',
           'Course name' => 'Not on Apply',
           'Provider code' => course.provider.code.to_s,
-          'Apply candidate ID' => candidate.id.to_s }
-      ucas_matching_application = UCASMatchedApplication.new(ucas_matching_data, recruitment_cycle_year)
+        }
+        ucas_matching_application = UCASMatchedApplication.new(ucas_matching_data, course.recruitment_cycle_year)
 
-      expect(ucas_matching_application.course.code).to eq('123')
-      expect(ucas_matching_application.course.name).to eq('Not on Apply')
-      expect(ucas_matching_application.course.provider).to eq(course.provider)
-    end
+        expect(ucas_matching_application.course.code).to eq('123')
+        expect(ucas_matching_application.course.name).to eq('Not on Apply')
+        expect(ucas_matching_application.course.provider).to eq(course.provider)
+      end
 
-    it 'handles missing course data' do
-      ucas_matching_data =
-        { 'Scheme' => 'U',
+      it 'handles missing course data' do
+        ucas_matching_data = {
+          'Scheme' => 'U',
           'Course code' => '',
           'Course name' => '',
           'Provider code' => 'T80',
-          'Apply candidate ID' => candidate.id.to_s }
-      ucas_matching_application = UCASMatchedApplication.new(ucas_matching_data, recruitment_cycle_year)
+        }
+        ucas_matching_application = UCASMatchedApplication.new(ucas_matching_data, recruitment_cycle_year)
 
-      expect(ucas_matching_application.course.code).to eq('Missing course code')
-      expect(ucas_matching_application.course.name).to eq('Missing course name')
-      expect(ucas_matching_application.course.provider).to eq(nil)
+        expect(ucas_matching_application.course.code).to eq('Missing course code')
+        expect(ucas_matching_application.course.name).to eq('Missing course name')
+        expect(ucas_matching_application.course.provider).to eq(nil)
+      end
     end
   end
 
   describe '#status' do
     context 'when it is a DfE match' do
+      let(:application_choice) { build_stubbed(:application_choice, course_option: course_option) }
+
       it 'returns the applications status' do
         dfe_matching_data =
           { 'Scheme' => 'D',
             'Course code' => course.code.to_s,
             'Provider code' => course.provider.code.to_s,
             'Apply candidate ID' => candidate.id.to_s }
         ucas_matching_application = UCASMatchedApplication.new(dfe_matching_data, recruitment_cycle_year)
+        allow(ucas_matching_application).to receive(:application_choice).and_return(application_choice)
```

Comment:
This looks like a hint that `#application_choice` should probably live somewhere else.

Rather than stub a method on the object under test, we could:

- Move the implementation of the `#application_choice` method to something appropriately named on the ApplicationChoice model.
- Stub that new method instead.
- Replace the testing we're doing for `#application_choice` in this spec file with something equivalent in `application_choice_spec.rb`.

Reply:
As the intention of this PR is to improve optimise the spec this is something we should address separately so I will add a card for it.